weasyl/message.py

```diff
@@ -69,27 +69,86 @@ def remove_all_submissions(userid, only_before=None):
     d._page_header_info.invalidate(userid)
 
 
-def select_journals(userid):
-    journals = d.engine.execute("""
-        SELECT we.welcomeid, we.unixtime, we.otherid, we.targetid, pr.username, jo.title
+def select_journals(userid, backtime=None, nexttime=None, limit=None, include_tags=False, include_content=False):
+    if limit:
+        limit_filter = "LIMIT %(limit)s"
+    else:
+        limit_filter = ""
+
+    if backtime:
+        time_filter = "AND we.unixtime > %(backtime)s"
+    elif nexttime:
+        time_filter = "AND we.unixtime < %(nexttime)s"
+    else:
+        time_filter = ""
+
+    if include_tags:
+        tags_select = ", COALESCE(array_agg(tagid) FILTER (WHERE tagid IS NOT NULL), '{}') AS tags"
+        tags_join = "LEFT JOIN searchmapjournal AS smj ON jo.journalid = smj.targetid"
+        tags_groupby = "GROUP BY jo.journalid, pr.username, we.welcomeid"
+    else:
+        tags_select = tags_join = tags_groupby = ""
+
+    if include_content:
+        content_select = ", jo.content"
+    else:
+        content_select = ""
+
+    statement = """
+        SELECT we.welcomeid, we.unixtime, we.otherid, jo.journalid, pr.username, jo.title, jo.rating {content_select} {tags_select}
         FROM welcome we
             INNER JOIN profile pr ON we.otherid = pr.userid
             INNER JOIN journal jo ON we.targetid = jo.journalid
+            {tags_join}
         WHERE
             (we.userid, we.type) = (%(user)s, 1010) AND
             rating <= %(rating)s
-        ORDER BY we.unixtime DESC
-    """, user=userid, rating=d.get_rating(userid))
+            {time_filter}
+        {tags_groupby}
+        ORDER BY we.welcomeid DESC
+        {limit_filter}
+        """.format(
```

Comment:
Minor formatting consistency with `select_submissions`:
```suggestion
    """.format(
```